src/client.cpp

```diff
@@ -160,25 +160,75 @@ void Client::initMods()
 
 	// Print mods
 	infostream << "Client Loading mods: ";
-	for (std::vector<ModSpec>::const_iterator i = mods.begin();
-		i != mods.end(); ++i) {
+	for (std::vector<ModSpec>::const_iterator i = m_mods.begin();
+		i != m_mods.end(); ++i) {
 		infostream << (*i).name << " ";
 	}
-
 	infostream << std::endl;
+
 	// Load and run "mod" scripts
-	for (std::vector<ModSpec>::const_iterator it = mods.begin();
-		it != mods.end(); ++it) {
+	for (std::vector<ModSpec>::const_iterator it = m_mods.begin();
+			it != m_mods.end(); ++it) {
 		const ModSpec &mod = *it;
 		if (!string_allowed(mod.name, MODNAME_ALLOWED_CHARS)) {
 			throw ModError("Error loading mod \"" + mod.name +
 				"\": Mod name does not follow naming conventions: "
 					"Only characters [a-z0-9_] are allowed.");
 		}
-		std::string script_path = mod.path + DIR_DELIM + "init.lua";
-		infostream << "  [" << padStringRight(mod.name, 12) << "] [\""
-			<< script_path << "\"]" << std::endl;
-		m_script->loadMod(script_path, mod.name);
+		loadModIntoMemory(mod.name, mod.path);
+	}
+}
+
+void Client::loadModIntoMemory(const std::string &mod_name, const std::string &mod_path)
+{
+	std::vector<fs::DirListNode> mod = fs::GetDirListing(mod_path);
+	for (unsigned int j=0; j < mod.size(); j++){
+		std::string filename = mod[j].name;
+		if (mod[j].dir) {
+			loadModSubfolder(mod_name, mod_path, filename + DIR_DELIM);
+			continue;
+		}
+		std::ifstream file (mod_path + DIR_DELIM + filename);
+		std::stringstream data;
+		data << file.rdbuf();
+		m_mod_files[mod_name + ":" + filename] = data.str();
+	}
+}
```

Comment:
I don't think this will actually make server-supplied mods much easier to implement -- there are only a dozen lines of relevant code, and it will probably need modification to work anyways.  Also, server-sent mods might be loaded the same way (they may be put in the on-disk cache and then loaded from there).